applications/ExaquteSandboxApplication/custom_processes/metrics_divergencefree_process.cpp

```diff
@@ -0,0 +1,385 @@
+    // Initialize to zero
+    for (int i_node = 0; i_node < number_nodes; ++i_node) {
+        divergence_nodal_values[i_node] = 0;
+        area_nodal_values[i_node] = 0;
+    }
+    // Weighted mean
+    const auto it_elem_begin = elements_array.begin();
+    for (int i_elem = 0; i_elem < number_elements; ++i_elem) {
+        auto it_elem = it_elem_begin + i_elem;
+        auto r_geometry = it_elem->GetGeometry();
+        for (unsigned int i_node = 0; i_node < r_geometry.size(); ++i_node) {
+            area_nodal_values[r_geometry[i_node].Id()-1] = area_nodal_values[r_geometry[i_node].Id()-1] + it_elem->GetGeometry().Area() / r_geometry.size();
+            divergence_nodal_values[r_geometry[i_node].Id()-1] = divergence_nodal_values[r_geometry[i_node].Id()-1] + it_elem->GetValue(r_reference_var) * it_elem->GetGeometry().Area() / r_geometry.size();
+        }
+    }
+    // Delete area contribution
+    for (int i_node = 0; i_node < number_nodes; ++i_node) {
+        divergence_nodal_values[i_node] = divergence_nodal_values[i_node] / area_nodal_values[i_node];
+    }
+
+    // Maximum strategy
+    if (mRefinementStrategy == MaximumStrategy) {
+
+        const auto it_nodes_begin = nodes_array.begin();
+        #pragma omp parallel for
+        for(int i_node = 0; i_node < number_nodes; ++i_node) {
+            auto it_node = it_nodes_begin + i_node;
+
+            // Retrieve variable
+            const double divergencefree_elem_value = divergence_nodal_values[i_node];
+
+            // Estimate element size
+            const double nodal_h = it_node->GetValue(NODAL_H);
+
+            // Check for zero values
+            if (mMaxStrategyRefinementCoefficient < std::numeric_limits<double>::epsilon())
+            {
+                KRATOS_WARNING("maximum_strategy") << "WARNING: Your variable refinement coefficient is near zero: " << mMaxStrategyRefinementCoefficient << std::endl;
+                KRATOS_WARNING("maximum_strategy") << "WARNING: Setting refinement coefficient to lower numeric limit: " << std::numeric_limits<double>::epsilon() << std::endl;
+                mMaxStrategyRefinementCoefficient = std::numeric_limits<double>::epsilon();
+            }
+
+            // Set element size
+            double element_size;
+            if (divergencefree_elem_value >= mMaxStrategyTargetRefinementCoefficient*mDivergenceFreeMaxValue) {
+                element_size = nodal_h/mMaxStrategyRefinementCoefficient;
+            }
+            else {
+                element_size = nodal_h;
+            }
+
+            // Check with max and min size
+            if (element_size < mMinSize) element_size = mMinSize;
+            if (element_size > mMaxSize) element_size = mMaxSize;
+
+            // Set metric
+            BoundedMatrix<double, TDim, TDim> metric_matrix = ZeroMatrix(TDim, TDim);
+            for(IndexType i = 0;i < TDim; ++i)
+                metric_matrix(i,i) = 1.0/std::pow(element_size, 2);
```

Comment:
Check element_size > 0?

Reply:
Why? Later I do the check with min and max size

Reply:
OK, any case there are more comments